File: frontend/lib/models/application.dart
import 'scheme.dart';

class ApplicationItem {
  final String id;
  final String trackingId;
  final String userId;
  final String schemeId;
  final String schemeName;
  final num requestedAmount;
  final String proposedBusiness;
  String status; // 'Submitted', 'Under Review', 'Approved', 'Rejected'
  final List<SchemeDocument> uploadedDocuments;
  final DateTime createdAt;
  final String remarks;

  ApplicationItem({
    required this.id,
    required this.trackingId,
    required this.userId,
    required this.schemeId,
    required this.schemeName,
    required this.requestedAmount,
    required this.proposedBusiness,
    required this.status,
    required this.uploadedDocuments,
    required this.createdAt,
    this.remarks = '',
  });

  factory ApplicationItem.fromJson(Map<String, dynamic> json) {
    var docList = <SchemeDocument>[];
    if (json['uploadedDocuments'] != null) {
      docList = (json['uploadedDocuments'] as List)
          .map((d) => SchemeDocument.fromJson(d is Map<String, dynamic> ? d : {'docName': d.toString()}))
          .toList();
    }

    return ApplicationItem(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      trackingId: json['trackingId'] ?? 'UDS-000000',
      userId: (json['userId'] ?? '').toString(),
      schemeId: (json['schemeId'] ?? '').toString(),
      schemeName: json['schemeName'] ?? 'Government Scheme',
      requestedAmount: json['requestedAmount'] ?? 500000,
      proposedBusiness: json['proposedBusiness'] ?? 'Micro Enterprise',
      status: json['status'] ?? 'Submitted',
      uploadedDocuments: docList,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      remarks: json['remarks'] ?? '',
    );
  }
}

File: backend/models/Application.js
const mongoose = require('mongoose');

const UploadedDocumentSchema = new mongoose.Schema({
  docName: {
    type: String,
    required: true
  },
  url: {
    type: String,
    default: ''
  },
  fileName: {
    type: String,
    default: ''
  },
  fileType: {
    type: String,
    default: 'image/jpeg'
  },
  fileSize: {
    type: String,
    default: '1.2 MB'
  },
  status: {
    type: String,
    enum: ['Uploaded', 'Pending', 'Verified', 'Rejected'],
    default: 'Pending'
  },
  uploadedAt: {
    type: Date,
    default: Date.now
  }
});

const ApplicationSchema = new mongoose.Schema({
  trackingId: {
    type: String,
    unique: true,
    required: true,
    default: () => 'UDS-' + Math.floor(100000 + Math.random() * 900000)
  },
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  schemeId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Scheme',
    required: true
  },
  requestedAmount: {
    type: Number,
    default: 500000
  },
  proposedBusiness: {
    type: String,
    default: 'Food Processing / Retail Venture'
  },
  status: {
    type: String,
    enum: ['Draft', 'Submitted', 'Under Review', 'Verified by Partner', 'Approved', 'Disbursed', 'Rejected'],
    default: 'Submitted'
  },
  assignedPartnerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'ChannelPartner'
  },
  uploadedDocuments: [UploadedDocumentSchema],
  remarks: {
    type: String,
    default: 'Application initiated via Udyam Setu mobile portal.'
  }
}, {
  timestamps: true
});

module.exports = mongoose.models.Application || mongoose.model('Application', ApplicationSchema);
